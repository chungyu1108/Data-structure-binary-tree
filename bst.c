/*
 * This file is where you should implement your binary search tree.  It already
 * contains skeletons of the functions you need to implement (along with
 * documentation for each function).  Feel free to implement any additional
 * functions you might need.  Also, don't forget to include your name and
 * @oregonstate.edu email address below.
 *
 * Name: Chung-Yu, Yang
 * Email: yangchun@oregonstate.edu
 */

#include <stdlib.h>

#include "bst.h"
#include "stack.h"
#include <assert.h>

/*
 * This structure represents a single node in a BST.  In addition to containing
 * pointers to its two child nodes (i.e. `left` and `right`), it contains two
 * fields representing the data stored at this node.  The `key` field is an
 * integer value that should be used as an identifier for the data in this
 * node.  Nodes in the BST should be ordered based on this `key` field.  The
 * `value` field stores data associated with the key.
 *
 * You should not modify this structure.
 */
struct bst_node {
  int key;
  void* value;
  struct bst_node* left;
  struct bst_node* right;
};


/*
 * This structure represents an entire BST.  It specifically contains a
 * reference to the root node of the tree.
 *
 * You should not modify this structure.
 */
struct bst {
  struct bst_node* root;
};

/*
 * This function should allocate and initialize a new, empty, BST and return
 * a pointer to it.
 */
struct bst* bst_create(int val) 
{
  struct bst* tree = malloc(sizeof(struct bst));
  tree->root = NULL;
  return tree;
}

/*
 * This function should free the memory associated with a BST.  While this
 * function should up all memory used in the BST itself, it should not free
 * any memory allocated to the pointer values stored in the BST.  This is the
 * responsibility of the caller.
 *
 * Params:
 *   bst - the BST to be destroyed.  May not be NULL.
 */
void free_bst_node(struct bst_node* node) 
{
  if (node->left == NULL && node->right == NULL) 
  {
    free(node);
  }
  else if (node->right != NULL) 
  {
    free_bst_node(node->right);
  }
  else (node->left != NULL);
  {
    free_bst_node(node->left);
  }
  free(node);
}

void bst_free(struct bst* bst) 
{
  free_bst_node(bst->root);
  free(bst);
  return;
}

/*
 * This function should return the total number of elements stored in a given
 * BST.
 *
 * Params:
 *   bst - the BST whose elements are to be counted.  May not be NULL.
 */

int bst_subtree(struct bst_node* root)
{
  int size = 1;
  //number nodes in right tree of root
  if(root->right != NULL)
  {
    size += bst_subtree(root->right);
  }
  //number nodes in left tree of root
  if(root->left != NULL)
  {
    size += bst_subtree(root->left);
  }
return size;
}

int bst_size(struct bst* bst) 
{
  assert(bst);
  int size = 0;
  if(bst->root != NULL)
  {
    size = bst_subtree(bst->root);
  }
  return size;
}

/*
 * This function should insert a new key/value pair into the BST.  The key
 * should be used to order the key/value pair with respect to the other data
 * stored in the BST.  The value should be stored along with the key, once the
 * right location in the tree is found.
 *
 * Params:
 *   bst - the BST into which a new key/value pair is to be inserted.  May not
 *     be NULL.
 *   key - an integer value that should be used to order the key/value pair
 *     being inserted with respect to the other data in the BST.
 *   value - the value being inserted into the BST.  This should be stored in
 *     the BST alongside the key.  Note that this parameter has type void*,
 *     which means that a pointer of any type can be passed.
 */
void bst_insert(struct bst* bst, int key, void* value) 
{
  struct bst_node* ptr;
  struct bst_node* tree = malloc(sizeof(struct bst_node));
    
  tree->key = key;
  tree->value = value;
  tree->right = NULL;
  tree->left = NULL;
  
  if(bst->root == NULL)
  {
    bst->root = tree;
    return;
  }
  else {
    ptr = bst->root;
  }
  //using the while to insert check the key for the left side 
  //and right side
  int node = 1;
  while(node == 1)
  {
    if(tree->key >= ptr->key)
    {
      if(ptr->right == NULL)
      {
        node = 0;
        ptr->right = tree;
      }
      else 
        ptr = ptr->right;
    }
    else if(tree->key < ptr->key)
    {
      if(ptr->left == NULL)
      {
        node = 0;
        ptr->left = tree;     
      }
      else 
        ptr = ptr->left;
    }
  }
  return;
}


/*
 * This function should remove a key/value pair with a specified key from a
 * given BST.  If multiple values with the same key exist in the tree, this
 * function should remove the first one it encounters (i.e. the one closest to
 * the root of the tree).
 *
 * Params:
 *   bst - the BST from which a key/value pair is to be removed.  May not
 *     be NULL.
 *   key - the key of the key/value pair to be removed from the BST.
 */
void bst_remove(struct bst* bst, int key) 
{
  struct bst_node* node_n = bst->root;
  struct bst_node* prve = NULL;

  while(node_n != NULL && key != node_n->key)
  {
    prve = node_n;
    if(key < node_n->key)
    {
      node_n = node_n->left;
    }
    else if (node_n->key < key)
    {
      node_n = node_n->right;
    }
    else break;
  }
  //Check if the node left and right is the NULL
  if(node_n->left == NULL && node_n->right == NULL)
  {
    if(node_n == prve->left)
    {
      prve->left = NULL;
    }
    else
      prve->right = NULL;
  }
  //Check if the left is not NULL and if right node is NULL
  else if(node_n->left != NULL && node_n->right == NULL)
  {
    if(node_n == prve->left)
    {
      prve->left = node_n->left;
    }
    else
      prve->right = node_n->left;
  }
  //Check if the right is not NULL and if left node is NULL
  else if(node_n->left == NULL && node_n->right != NULL)
  {
    if(node_n == prve->left)
    {
      prve->left = node_n->right;
    }
    else
      prve->right = node_n->right;
  }
  
  else {
    struct bst_node* node_s;
    struct bst_node* parent_s;

    node_s = node_n->right;
    parent_s = node_n;
    while(node_s->left != NULL && node_s != NULL)
    {
      parent_s = node_s;
      node_s = node_s->left;
    }
      node_s->left = node_n->left;
      if(node_s != node_n->right)
      {
        parent_s->left = node_s->right;
        node_s->right = node_n->right;
      }
      if (prve == NULL)
      {
        bst->root = node_s;
      }
      else if(node_n == prve->left)
      {
        prve->left = node_s;
      }
      else
        prve->right = node_s;
  }
  free(node_n);
  return;
}
/*
 * This function should return the value associated with a specified key in a
 * given BST.  If multiple values with the same key exist in the tree, this
 * function should return the first one it encounters (i.e. the one closest to
 * the root of the tree).  If the BST does not contain the specified key, this
 * function should return NULL.
 *
 * Params:
 *   bst - the BST from which a key/value pair is to be removed.  May not
 *     be NULL.
 *   key - the key of the key/value pair whose value is to be returned.
 *
 * Return:
 *   Should return the value associated with the key `key` in `bst` or NULL,
 *   if the key `key` was not found in `bst`.
 */

void* get_bst_node(struct bst_node *ptr, int key) 
{
  if(ptr == NULL) 
    return NULL;

  else if(ptr->key == key) 
    return ptr->value;

  else if(key < ptr->key) 
    return get_bst_node(ptr->left, key);

  else 
    return get_bst_node(ptr->right, key);
}

void* bst_get(struct bst* bst, int key) 
{
  if(bst == NULL)
    return NULL;

  return get_bst_node(bst->root, key);
}


/*****************************************************************************
 **
 ** BST puzzle functions
 **
 *****************************************************************************/

/*
 * This function should return the height of a given BST, which is the maximum
 * depth of any node in the tree (i.e. the number of edges in the path from
 * the root to that node).  Note that the height of an empty tree is -1 by
 * convention.
 *
 * Params:
 *   bst - the BST whose height is to be computed
 *
 * Return:
 *   Should return the height of bst.
 */
int bst_height_val(struct bst_node* root)
{
  if(root == NULL)
  {
    return -1;
  }
  int left = bst_height_val(root->left);
  int right = bst_height_val(root->right);
    return left > right ? left +1 : right +1;
}

 int bst_height(struct bst* bst) 
 {
  return bst_height_val(bst->root);
 }

/*
 * This function should determine whether a specified value is a valid path
 * sum within a given BST.  In other words, this function should check whether
 * the given BST contains any path from the root to a leaf in which the keys
 * sum to the specified value.
 *
 * Params:
 *   bst - the BST whose paths sums to search
 *   sum - the value to search for among the path sums of `bst`
 *
 * Return:
 *   Should return 1 if `bst` contains any path from the root to a leaf in
 *   which the keys add up to `sum`.  Should return 0 otherwise.
 */
int bst_path_sum(struct bst* bst, int sum) {
  return 0;//bst_path_sum_tree(sum, bst->root);
}

/*int bst_path_sum_tree(int sum, struct bst_node* ptr)
{
  if(ptr == NULL)
    return 0;
    
  if(ptr->left == NULL && ptr->right == NULL)
    return 1;
  int left = bst_path_sum_tree(sum, ptr->left);
  int right = bst_path_sum_tree(sum, ptr-right);
    return left || right;
}*/

/*
 * This function should compute a range sum in a given BST.  Specifically, it
 * should compute the sum of all keys in the BST between a given lower bound
 * and a given upper bound.  For full credit, you should not process any subtree
 * whose keys cannot be included in the range sum.
 *
 * Params:
 *   bst - the BST within which to compute a range sum
 *   lower - the lower bound of the range over which to compute a sum; this
 *     should be considered an *inclusive* bound; in other words a key that's
 *     equal to this bound should be included in the sum
 *   upper - the upper bound of the range over which to compute a sum; this
 *     should be considered an *inclusive* bound; in other words a key that's
 *     equal to this bound should be included in the sum
 *
 * Return:
 *   Should return the sum of all keys in `bst` between `lower` and `upper`.
 */

int get_bst_range_sum(struct bst_node* ptr, int lower, int upper)
{
  if (ptr == NULL) {
    return 0;
  }
    else if(ptr->key > upper)
    {
      return get_bst_range_sum(ptr->left, lower, upper);
    }
    else if(ptr->key < lower)
    {
      return get_bst_range_sum(ptr->right, lower, upper);
    }
  else
    return ptr->key + get_bst_range_sum(ptr->left, lower, upper) + get_bst_range_sum(ptr->right, lower, upper);
}

int bst_range_sum(struct bst* bst, int lower, int upper) 
{
  return get_bst_range_sum(bst->root, lower, upper);
}

/*****************************************************************************
 **
 ** BST iterator definition (extra credit only)
 **
 *****************************************************************************/

/*
 * Structure used to represent a binary search tree iterator.  It contains
 * only a reference to a stack to be used to implement the iterator.
 *
 * You should not modify this structure.
 */
struct bst_iterator {
  struct stack* stack;
};

/*
 * This function should allocate and initialize an iterator over a specified
 * BST and return a pointer to that iterator.
 *
 * Params:
 *   bst - the BST for over which to create an iterator.  May not be NULL.
 */
struct bst_iterator* bst_iterator_create(struct bst* bst) {
  return NULL;
}

/*
 * This function should free all memory allocated to a given BST iterator.
 * It should NOT free any memory associated with the BST itself.  This is the
 * responsibility of the caller.
 *
 * Params:
 *   iter - the BST iterator to be destroyed.  May not be NULL.
 */
void bst_iterator_free(struct bst_iterator* iter) {
  return;
}

/*
 * This function should indicate whether a given BST iterator has more nodes
 * to visit.  It should specifically return 1 (true) if the iterator has at
 * least one more node to visit or 0 (false) if it does not have any more
 * nodes to visit.
 *
 * Param:
 *   iter - the BST iterator to be checked for remaining nodes to visit.  May
 *     not be NULL.
 */
int bst_iterator_has_next(struct bst_iterator* iter) {
  return 0;
}

/*
 * This function should return both the value and key associated with the
 * current node pointed to by the specified BST iterator and advnce the
 * iterator to point to the next node in the BST (in in-order order).
 *
 * Because a function can't return two things, the key associated with the
 * current node should be returned the normal way, while its value should be
 * returned via the argument `value`.  Specifically, the argument `value`
 * is a pointer to a void pointer.  The current BST node's value (a void
 * pointer) should be stored at the address represented by `value` (i.e. by
 * dereferencing `value`).  This will look something like this:
 *
 *   *value = current_node->value;
 *
 * Parameters:
 *   iter - BST iterator.  The key and value associated with this iterator's
 *     current node should be returned, and the iterator should be updated to
 *     point to the next node in the BST (in in-order order).  May not be NULL.
 *   value - pointer at which the current BST node's value should be stored
 *     before this function returns.
 *
 * Return:
 *   This function should return the key associated with the current BST node
 *   pointed to by `iter`.
 */
int bst_iterator_next(struct bst_iterator* iter, void** value) {
  if (value) {
    *value = NULL;
  }
  return 0;
}
