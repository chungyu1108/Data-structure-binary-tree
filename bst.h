/*
 * This file contains the definition of the interface for the binary search
 * tree you'll implement.  You should not modify anything in this file.  You
 * can find descriptions of the binary search tree functions, including their
 * parameters and their return values, in bst.c.
 */

#ifndef __BST_H
#define __BST_H

/*
 * Structure used to represent a binary search tree.
 */
struct bst;

/*
 * Basic binary search tree interface function prototypes.  Refer to bst.c for
 * documentation about each of these functions.
 */
struct bst* bst_create();
void bst_free(struct bst* bst);
int bst_size(struct bst* bst);
void bst_insert(struct bst* bst, int key, void* value);
void bst_remove(struct bst* bst, int key);
void* bst_get(struct bst* bst, int key);

/*
 * Binary search tree "puzzle" function prototypes.  Refer to bst.c for
 * documentation about each of these functions.
 */
int bst_height(struct bst* bst);
int bst_path_sum(struct bst* bst, int sum);
//int bst_path_sum_tree(int cur, int sum, struct bst_node* root);
int bst_range_sum(struct bst* bst, int lower, int upper);

/*
 * Structure used to represent a binary search tree iterator.
 */
struct bst_iterator;

/*
 * Binary search tree iterator interface prototypes.  Refer to bst.c for
 * documentation about each of these functions.
 */
struct bst_iterator* bst_iterator_create(struct bst* bst);
void bst_iterator_free(struct bst_iterator* iter);
int bst_iterator_has_next(struct bst_iterator* iter);
int bst_iterator_next(struct bst_iterator* iter, void** value);

#endif
