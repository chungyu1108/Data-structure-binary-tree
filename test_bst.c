/*
 * This file contains executable code for testing your BST implementation.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "bst.h"

/*
 * This is the data that's used to test this program.  It forms a tree that
 * looks like this:
 *
 *               64
 *              /  \
 *             /    \
 *            /      \
 *           /        \
 *          32        96
 *         /  \      /  \
 *        /    \    /    \
 *       16    48  80    112
 *      /  \     \   \   /  \
 *     8   24    56  88 104 120
 */
#define NUM_TEST_DATA 13
const int TEST_DATA[NUM_TEST_DATA] =
  {64, 32, 96, 16, 48, 80, 112, 8, 24, 56, 88, 104, 120};

/*
 * This is the correct height of the tree defined above.
 */
#define TEST_DATA_BST_HEIGHT 3

/*
 * This array contains all of the valid path sums in the tree defined above.
 */
#define NUM_GOOD_PATH_SUMS 6
const int GOOD_PATH_SUMS[NUM_GOOD_PATH_SUMS] = {120, 136, 200, 328, 376, 392};

/*
 * This array defines some range sums to compute within the tree defined above.
 * Each range sum is specified as a triple: {lower, upper, sum}, where `lower`
 * and `upper` are respectively the lower and upper bounds of the range and
 * `sum` is the true sum of keys in that range.
 */
#define NUM_RANGE_SUMS 10
const int RANGE_SUMS[NUM_RANGE_SUMS][3] = {
  {8, 120, 848},
  {0, 200, 848},
  {2, 40, 80},
  {24, 60, 160},
  {30, 90, 368},
  {60, 70, 64},
  {60, 112, 544},
  {84, 110, 288},
  {96, 96, 96},
  {125, 200, 0}
};

/*
 * This array contains values from the TEST_DATA array above that we'll try
 * to remove from the BST.
 */
#define NUM_DATA_TO_REMOVE 4
const int TEST_DATA_TO_REMOVE[NUM_DATA_TO_REMOVE] = {16, 48, 64, 104};

/*
 * This is a helper function that's used to compare integers when sorting with
 * qsort().
 */
int cmp_ints(const void* a, const void* b) {
  return *(int*)a - *(int*)b;
}

int main(int argc, char** argv) {
  /*
   * Create a new BST and insert the testing data into it.  The testing data
   * is inserted in such a way that the value is equal to the address of the
   * corresponding key (which comes from the TEST_DATA array above).  In other
   * words, for any given key, we'll be able to verify that the tree contains
   * the correct value by comparing key == *value.
   */
  printf("== Creating BST...\n");
  struct bst* bst = bst_create();
  printf("\n== Inserting %d values into BST...\n", NUM_TEST_DATA);
  for (int i = 0; i < NUM_TEST_DATA; i++) {
    bst_insert(bst, TEST_DATA[i], (void*)&TEST_DATA[i]);
  }

  /*
   * Make sure bst_size() returns the correct value.
   */
  printf("\n== Checking correct value from bst_size(): %d (expected %d)\n",
    bst_size(bst), NUM_TEST_DATA);

  /*
   * Make sure bst_size() returns the correct value.
   */
  printf("\n== Checking correct value from bst_height(): %d (expected %d)\n",
    bst_height(bst), TEST_DATA_BST_HEIGHT);

  /*
   * Test that we can successfully lookup all of the values we know are in the
   * BST.
   */
  printf("\n== Looking up values we know should be in the BST...\n");
  for (int i = 0; i < NUM_TEST_DATA; i++) {
    int* value = bst_get(bst, TEST_DATA[i]);
    if (value) {
      printf("  -- bst_get(%3d): %3d (expected %3d)\n", TEST_DATA[i], *value,
        TEST_DATA[i]);
    } else {
      printf("  -- bst_get(%3d) unexpectedly returned NULL\n", TEST_DATA[i]);
    }
  }

  /*
   * Test that we *can't* successfully find any values we know shouldn't be in
   * the BST.  To do this, create a sorted copy of the test data that will
   * allow us to more easily determine which values shouldn't be in the BST.
   */
  int* sorted = malloc(NUM_TEST_DATA * sizeof(int));
  memcpy(sorted, TEST_DATA, NUM_TEST_DATA * sizeof(int));
  qsort(sorted, NUM_TEST_DATA, sizeof(int), cmp_ints);
  printf("\n== Looking up values we know should NOT be in the BST...\n");
  for (int i = 0, k = 0; i < sorted[NUM_TEST_DATA - 1]; i++) {
    if (i == sorted[k]) {
      /*
       * If i is equal to a key we know is in the BST, skip it.
       */
      k++;
    } else if (bst_get(bst, i) != NULL) {
      printf("  -- bst_get(%3d) unexpectedly returned a non-NULL value\n", i);
    }
  }

  /*
   * Test BST path sum function, checking several values that the tree should
   * and have as path sums.
   */
  printf("\n== Checking path sums the BST should contain:\n");
  for (int i = 0; i < NUM_GOOD_PATH_SUMS; i++) {
    printf("  == %3d: ", GOOD_PATH_SUMS[i]);
    if (bst_path_sum(bst, GOOD_PATH_SUMS[i])) {
      printf("all good\n");
    } else {
      printf("bst_path_sum() says NO when it should say YES\n");
    }
  }

  /*
   * Make sure that no invalid path sums are erroneously reported as valid.
   */
  printf("\n== Checking path sums the BST should NOT contain:\n");
  int num_bad_sums = 0;
  int cur_good_sum = 0;
  for (int i = 0; i < GOOD_PATH_SUMS[NUM_GOOD_PATH_SUMS - 1]; i += 4) {
    if (i == GOOD_PATH_SUMS[cur_good_sum]) {
      cur_good_sum++;
    } else if (bst_path_sum(bst, i)) {
      num_bad_sums++;
      printf("  -- %3d: bst_path_sum() says YES when it should say NO\n", i);
    }
  }
  printf("  -- found %d paths sums we shouldn't have found\n", num_bad_sums);

  /*
   * Test computation of range sums in BST.
   */
  printf("\n== Checking range sums in the BST:\n");
  for (int i = 0; i < NUM_RANGE_SUMS; i++) {
    int lower = RANGE_SUMS[i][0];
    int upper = RANGE_SUMS[i][1];
    int sum = RANGE_SUMS[i][2];
    printf("  -- bst_range_sum(%d, %d): %d (expected %d)\n", lower, upper,
      bst_range_sum(bst, lower, upper), sum);
  }


  /*
   * Test removing keys from the BST.  After removing each key, make sure
   * it's no longer present in the BST.
   */
  printf("\n== Removing keys from BST...\n");
  for (int i = 0; i < NUM_DATA_TO_REMOVE; i++) {
    bst_remove(bst, TEST_DATA_TO_REMOVE[i]);
    int* value = bst_get(bst, TEST_DATA_TO_REMOVE[i]);
    if (value) {
      printf("  -- key %3d still present in BST after removal, value: %d\n",
        TEST_DATA_TO_REMOVE[i], *value);
    } else {
      printf("  -- key %3d correctly removed from BST\n",
        TEST_DATA_TO_REMOVE[i]);
    }
  }

  /*
   * Make sure bst_size() still returns the correct value.
   */
  printf("\n== Checking correct value from bst_size(): %d (expected %d)\n",
    bst_size(bst), NUM_TEST_DATA - NUM_DATA_TO_REMOVE);

  /*
   * Test that all of the values we know should still be in the BST are indeed
   * there.
   */
  printf("\n== Looking up values we know should still be in the BST...\n");
  for (int i = 0, k = 0; i < NUM_TEST_DATA; i++) {
    /*
     * If the current key is one we've already removed, skip over it.
     */
    if (sorted[i] == TEST_DATA_TO_REMOVE[k]) {
      k++;
      continue;
    }

    int* value = bst_get(bst, sorted[i]);
    if (value) {
      printf("  -- bst_get(%3d): %3d (expected %3d)\n", sorted[i], *value,
        sorted[i]);
    } else {
      printf("  -- bst_get(%3d) unexpectedly returned NULL\n", sorted[i]);
    }
  }

  free(sorted);
  bst_free(bst);

  return 0;
}
