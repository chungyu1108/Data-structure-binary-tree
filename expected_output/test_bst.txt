$ ./test_bst
== Creating BST...

== Inserting 13 values into BST...

== Checking correct value from bst_size(): 13 (expected 13)

== Checking correct value from bst_height(): 3 (expected 3)

== Looking up values we know should be in the BST...
  -- bst_get( 64):  64 (expected  64)
  -- bst_get( 32):  32 (expected  32)
  -- bst_get( 96):  96 (expected  96)
  -- bst_get( 16):  16 (expected  16)
  -- bst_get( 48):  48 (expected  48)
  -- bst_get( 80):  80 (expected  80)
  -- bst_get(112): 112 (expected 112)
  -- bst_get(  8):   8 (expected   8)
  -- bst_get( 24):  24 (expected  24)
  -- bst_get( 56):  56 (expected  56)
  -- bst_get( 88):  88 (expected  88)
  -- bst_get(104): 104 (expected 104)
  -- bst_get(120): 120 (expected 120)

== Looking up values we know should NOT be in the BST...

== Checking path sums the BST should contain:
  == 120: all good
  == 136: all good
  == 200: all good
  == 328: all good
  == 376: all good
  == 392: all good

== Checking path sums the BST should NOT contain:
  -- found 0 paths sums we shouldn't have found

== Checking range sums in the BST:
  -- bst_range_sum(8, 120): 848 (expected 848)
  -- bst_range_sum(0, 200): 848 (expected 848)
  -- bst_range_sum(2, 40): 80 (expected 80)
  -- bst_range_sum(24, 60): 160 (expected 160)
  -- bst_range_sum(30, 90): 368 (expected 368)
  -- bst_range_sum(60, 70): 64 (expected 64)
  -- bst_range_sum(60, 112): 544 (expected 544)
  -- bst_range_sum(84, 110): 288 (expected 288)
  -- bst_range_sum(96, 96): 96 (expected 96)
  -- bst_range_sum(125, 200): 0 (expected 0)

== Removing keys from BST...
  -- key  16 correctly removed from BST
  -- key  48 correctly removed from BST
  -- key  64 correctly removed from BST
  -- key 104 correctly removed from BST

== Checking correct value from bst_size(): 9 (expected 9)

== Looking up values we know should still be in the BST...
  -- bst_get(  8):   8 (expected   8)
  -- bst_get( 24):  24 (expected  24)
  -- bst_get( 32):  32 (expected  32)
  -- bst_get( 56):  56 (expected  56)
  -- bst_get( 80):  80 (expected  80)
  -- bst_get( 88):  88 (expected  88)
  -- bst_get( 96):  96 (expected  96)
  -- bst_get(112): 112 (expected 112)
  -- bst_get(120): 120 (expected 120)
